### ISSUE TITLE
jemalloc upgrade needed to fix macOS Sierra

### ISSUE THREAD

**Issue by ilovezfs** [ablab/spades#9]
The vendored version of jemalloc in SPAdes is outdated and does not include the upstream jemalloc fixes for macOS Sierra (10.12). As a result, SPAdes is currently quite broken on Sierra.

The exception displayed is

```
malloc: *** malloc_zone_unregister() failed for 0x7fffb1c3c000
```
and then it hangs indefinitely.

Luckily, this has been fixed upstream in https://github.com/jemalloc/jemalloc/pull/427 and the fix is included in the latest release: https://github.com/jemalloc/jemalloc/releases

So it would be great if you could upgrade the vendored jemalloc so that SPAdes works on macOS 10.12.

A minimal patch to SPAdes is here: https://gist.githubusercontent.com/ilovezfs/931c240e527c55643fd72cb6df87db50/raw/22fcf917a48f434cebdbefe47cd9dd4a78fb64d4/gistfile1.txt

However, I would highly recommend using jemalloc 4.3.1 and not using the minimal patch.

Here is a full log of what the failure looks like:
```
bash-4.4$ spades.py --test
Command line: /usr/local/bin/spades.py	--test	

System information:
  SPAdes version: 3.9.0
  Python version: 2.7.12
  OS: Darwin-16.1.0-x86_64-i386-64bit

Output dir: /usr/local/Homebrew/Library/Taps/homebrew/homebrew-science/spades_test
Mode: read error correction and assembling
Debug mode is turned OFF

Dataset parameters:
  Multi-cell mode (you should set '--sc' flag if input data was obtained with MDA (single-cell) technology or --meta flag if processing metagenomic dataset)
  Reads:
    Library number: 1, library type: paired-end
      orientation: fr
      left reads: ['/usr/local/Cellar/spades/3.9.0/share/spades/test_dataset/ecoli_1K_1.fq.gz']
      right reads: ['/usr/local/Cellar/spades/3.9.0/share/spades/test_dataset/ecoli_1K_2.fq.gz']
      interlaced reads: not specified
      single reads: not specified
Read error correction parameters:
  Iterations: 1
  PHRED offset will be auto-detected
  Corrected reads will be compressed (with gzip)
Assembly parameters:
  k: automatic selection based on read length
  Repeat resolution is enabled
  Mismatch careful mode is turned OFF
  MismatchCorrector will be SKIPPED
  Coverage cutoff is turned OFF
Other parameters:
  Dir for temp files: /usr/local/Homebrew/Library/Taps/homebrew/homebrew-science/spades_test/tmp
  Threads: 16
  Memory limit (in Gb): 250


======= SPAdes pipeline started. Log can be found here: /usr/local/Homebrew/Library/Taps/homebrew/homebrew-science/spades_test/spades.log


===== Read error correction started. 


== Running read error correction tool: /usr/local/Cellar/spades/3.9.0/bin/hammer /usr/local/Homebrew/Library/Taps/homebrew/homebrew-science/spades_test/corrected/configs/config.info

hammer(13447,0x7fffb1c463c0) malloc: *** malloc_zone_unregister() failed for 0x7fffb1c3c000
```

**Comment by asl** [comment]
Yes, we are aware about the issue and working on porting the latest jemalloc to SPAdes. We had some local changes and therefore we're investigating whether they are necessary at all. Also jemalloc API slightly changed so we need to adopt that as well.

**Comment by ilovezfs** [comment]
It might be worth using the minimal patch above in a hot fix release then.

**Comment by asl** [comment]
Right. But we're almost here with the "proper" release :)

**Comment by ilovezfs** [comment]
Ah ok. No worries. I just pushed the patched version on Homebrew
https://github.com/Homebrew/homebrew-science/commit/acf0e5b0da36e9bd754f2c8c813f284dad6720b4
https://github.com/Homebrew/homebrew-science/commit/793132b9496218bf22b640288509729fe016ea0b

**Comment by asl** [comment]
Fixed in 3.9.1 and later on

**Comment by ilovezfs** [comment]
@asl Thanks! https://github.com/Homebrew/homebrew-science/pull/4957.
